proc/fuse/fuse.go

```diff
@@ -2,65 +2,155 @@ package fuse
 
 import (
 	"errors"
+	"sync"
 
 	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/proc/spill"
 	"github.com/brimsec/zq/zbuf"
 	"github.com/brimsec/zq/zcode"
 	"github.com/brimsec/zq/zng"
 	"github.com/brimsec/zq/zng/resolver"
 )
 
 var MemMaxBytes = 128 * 1024 * 1024
+var BatchSize = 100
 
 type Proc struct {
-	pctx   *proc.Context
-	parent proc.Interface
-	recs   []*zng.Record
-	done   bool
+	pctx     *proc.Context
+	parent   proc.Interface
+	spiller  *spill.File
+	slotByID [][]int
+	slots    []slot
+	uberType *zng.TypeRecord
+	once     sync.Once
+	resultCh chan proc.Result
+	nbytes   int
+	recs     []*zng.Record
+	types    resolver.Slice
 }
 
 func New(pctx *proc.Context, parent proc.Interface) (*Proc, error) {
 	return &Proc{
-		pctx:   pctx,
-		parent: parent,
+		pctx:     pctx,
+		parent:   parent,
+		resultCh: make(chan proc.Result),
 	}, nil
 }
 
 func (p *Proc) Pull() (zbuf.Batch, error) {
-	if p.done {
-		return nil, nil
+	p.once.Do(func() { go p.run() })
+	if r, ok := <-p.resultCh; ok {
+		return r.Batch, r.Err
+	}
+	return nil, p.pctx.Err()
+}
+
+func (p *Proc) run() {
+	if err := p.pullInput(); err != nil {
+		p.shutdown(err)
+		return
 	}
-	var nbytes int
-	var types resolver.Slice
+	p.shutdown(p.pushOutput())
+}
+
+func (p *Proc) pullInput() error {
 	for {
+		if err := p.pctx.Err(); err != nil {
+			return err
+		}
 		batch, err := p.parent.Pull()
 		if err != nil {
-			return nil, err
+			return err
 		}
 		if batch == nil {
-			p.done = true
-			return p.finish(types, p.recs)
-		}
-		l := batch.Length()
-		for i := 0; i < l; i++ {
-			rec := batch.Index(i)
-			id := rec.Type.ID()
-			typ := types.Lookup(id)
-			if typ == nil {
-				types.Enter(id, rec.Type)
+			return p.finish()
+		}
+		if err := p.writeBatch(batch); err != nil {
+			return err
+		}
+	}
+}
+
+func (p *Proc) writeBatch(batch zbuf.Batch) error {
+	l := batch.Length()
```

Comment:
Should we add a `defer batch.Unref()` here?

Reply:
Good point.  This entailed adding a bit of logic to stash to do the Keep when stashing and not spilling.